composer/workflows/s3togcsoperator_tutorial.py

```diff
@@ -0,0 +1,149 @@
+)
+from airflow.providers.google.cloud.transfers.s3_to_gcs import S3ToGCSOperator
+from airflow.utils.task_group import TaskGroup
+
+PROJECT_NAME = "{{var.value.gcp_project}}"
+
+# BigQuery configs
+BQ_DESTINATION_DATASET_NAME = "holiday_weather"
+BQ_DESTINATION_TABLE_NAME = "holidays_weather_joined"
+BQ_NORMALIZED_TABLE_NAME = "holidays_weather_normalized"
+
+# Dataproc configs
+BUCKET_NAME = "{{var.value.gcs_bucket}}"
+PYSPARK_JAR = "gs://spark-lib/bigquery/spark-bigquery-latest_2.12.jar"
+PROCESSING_PYTHON_FILE = f"gs://{BUCKET_NAME}/data_analytics_process.py"
+
+# S3 configs
+S3_BUCKET_NAME = "{{var.value.s3_bucket}}"
+
+BATCH_ID = "data-processing-{{ ts_nodash | lower}}"  # Dataproc serverless only allows lowercase characters
+BATCH_CONFIG = {
+    "pyspark_batch": {
+        "jar_file_uris": [PYSPARK_JAR],
+        "main_python_file_uri": PROCESSING_PYTHON_FILE,
+        "args": [
+            BUCKET_NAME,
+            f"{BQ_DESTINATION_DATASET_NAME}.{BQ_DESTINATION_TABLE_NAME}",
+            f"{BQ_DESTINATION_DATASET_NAME}.{BQ_NORMALIZED_TABLE_NAME}",
+        ],
+
+    },
+    "environment_config": {
+        "execution_config": {
+            "service_account": "{{var.value.dataproc_service_account}}"
+        }
+    }
+}
+
+yesterday = datetime.datetime.combine(
+    datetime.datetime.today() - datetime.timedelta(1), datetime.datetime.min.time()
+)
+
+default_dag_args = {
+    # Setting start date as yesterday starts the DAG immediately when it is
+    # detected in the Cloud Storage bucket.
+    "start_date": yesterday,
+    # To email on failure or retry set 'email' arg to your email and enable
+    # emailing here.
+    "email_on_failure": False,
+    "email_on_retry": False
+}
+
+with models.DAG(
+    "s3_to_gcs_dag",
+    # Continue to run DAG once per day
+    schedule_interval=datetime.timedelta(days=1),
+    default_args=default_dag_args,
+) as dag:
+
+    s3_to_gcs_op = S3ToGCSOperator(
+        task_id="s3_to_gcs",
+        bucket=S3_BUCKET_NAME,
+        gcp_conn_id="google_cloud_default",
+        aws_conn_id="aws_s3_connection",
+        dest_gcs=f"gs://{BUCKET_NAME}",
+    )
+
+    create_batch = dataproc.DataprocCreateBatchOperator(
+        task_id="create_batch",
+        project_id=PROJECT_NAME,
+        region="{{ var.value.gce_region }}",
+        batch=BATCH_CONFIG,
+        batch_id=BATCH_ID,
+    )
+
+    load_external_dataset = GCSToBigQueryOperator(
+        task_id="run_bq_external_ingestion",
+        bucket=BUCKET_NAME,
+        source_objects=["holidays.csv"],
+        destination_project_dataset_table=f"{BQ_DESTINATION_DATASET_NAME}.holidays",
+        source_format="CSV",
+        schema_fields=[
+            {"name": "Date", "type": "DATE"},
+            {"name": "Holiday", "type": "STRING"},
+        ],
+        skip_leading_rows=1,
+        write_disposition="WRITE_TRUNCATE"
+    )
+
+    with TaskGroup("join_bq_datasets") as bq_join_group:
+
+        for year in range(1997, 2022):
```

Comment:
Any reason as to why `1997` or `2022` is special? Naming these as constants and/or comments should help clear this:

```
YEAR_TO_START = 1997
CURRENT_YEAR = datetime.datetime.now().year

...

# Run the dataset from the predefined years to current year.
for year in range(YEAR_TO_START, CURRENT_YEAR):
```

Reply:
These are the only dates found in one of the datasets which is why we chose them - they will stay fixed instead of the current year. This is also identical to [this line](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/f2b85fb2e5fabcdfc46c2e7fd407db364f7acdde/composer/2022_airflow_summit/data_analytics_dag.py#L104) - my original intent was to not change anything that was identical to code that had already been approved @dandhlee, would you want to see this constant change in both files?

Reply:
Hmm I can see the benefit of running this for fixed year. Let's stick with the proposed set years.